### PR TITLE
Bugfix/rclpy mem leak

### PR DESCRIPTION
**Pull request by srsidd**
In reference to #74 .

### CODE REVIEW

File: rclpy/rclpy/__init__.py [hunk 1]
```diff
@@ -109,6 +110,7 @@ def spin_once(node, *, timeout_sec=None):
         if request:
             response = srv.callback(request, srv.srv_type.Response())
             srv.send_response(response, header)
+    gc.collect()
```
Comment (dirk-thomas) [inline rclpy/rclpy/__init__.py]: I don't think that we should explicitly call `gc.collect` - especially not in a function like `spinOnce` which is called in a potentially very high frequency.

Comment (wjwwood) [inline rclpy/rclpy/__init__.py]: Yeah, sorry if I miscommunicated, but we don't want to do this.

From the other thread I was just saying that we should definitely have the destructor call the fini function in C, but we should let the gc run when it feels like it. Even though that might throw valgrind off.

The rest of the patch can stay though I think (still haven't reviewed it yet).

Comment (srsidd) [inline rclpy/rclpy/__init__.py]: I tried doing that, but it did not seem to have any effect. Is there another way of achieving the same effect?

Comment (dirk-thomas) [inline rclpy/rclpy/__init__.py]: Maybe the tests can either trigger the GC before finishing or monkey patch rclpy to perform this on shutdown?

**Comment by yamokosk** [comment]
Here's my suggestion for fixing this:

```cpp
/// waitset destructor
void
rclpy_destroy_wait_set(PyObject* pywait_set)
{
  rcl_wait_set_t * wait_set = (rcl_wait_set_t *)PyCapsule_GetPointer(pywait_set, NULL);
  rcl_ret_t ret = rcl_wait_set_fini(wait_set);
  if (ret != RCL_RET_OK) {
    PyErr_Format(PyExc_RuntimeError,
      "Failed to destroy wait set: %s", rcl_get_error_string_safe());
  }
  PyMem_Free(wait_set);
  return;
}

/// Return a Capsule pointing to a zero initialized rcl_wait_set_t structure
static PyObject *
rclpy_get_zero_initialized_wait_set(PyObject * Py_UNUSED(self), PyObject * Py_UNUSED(args))
{
  rcl_wait_set_t * wait_set = (rcl_wait_set_t *)PyMem_Malloc(sizeof(rcl_wait_set_t));
  *wait_set = rcl_get_zero_initialized_wait_set();

  PyObject * pywait_set = PyCapsule_New(wait_set, NULL, rclpy_destroy_wait_set);

  return pywait_set;
}
```

In testing on my machine, this eliminated the leak that grew linearly with the number of calls to spin_once (and confirmed with valgrind that it was coming from wait_set). The problem is that this is not a universal fix for all the other `PyMem_Malloc` calls in `_rclpy.c`. It addresses a lot of them but for things like `publisher` that take two args for its `_fini` function, this pattern won't work.

```cpp
rcl_ret_t
rcl_publisher_fini(rcl_publisher_t * publisher, rcl_node_t * node)
```

**Comment by srsidd** [comment]
Closing pull request since this is not the right way of doing this. And there is an issue tracking this at #74 .

**Comment by BrannonKing** [comment]
Reincorporating the above rclpy_destroy_wait_set on a current build does not eliminate the memory leak.

**Comment by mikaelarguedas** [comment]
thanks @BrannonKing for following up on this

Yeah I think the main leak is elsewhere, especially given than the bigger the messages the bigger the leak (would go along this https://github.com/ros2/rclpy/issues/74#issuecomment-320333788 about PyMem_Malloc)
I'm planning on chasing it down once the refactor of rclpy to [use executors](https://github.com/ros2/rclpy/pull/104) has been merged.

For reference: can you give more detail about the code you're running and if you also observe leaks proportional to the amount of data transmitted? 

Feel free to comment on the open ticket #74 or to open another one if the leak doesn't happen in spin_once, otherwise we may lose track of comments on an already closed ticket.
Also any contribution reducing (even partially) the current leaks is more than welcome!

**Comment by BrannonKing** [comment]
I'm aware of #74 -- it is the exact bug I was looking at. I'm not sending any messages (in my test). spin_once with a small/zero timeout leaks like the Titanic. It's super easy to reproduce: make a node and call spin_once repeatedly.

**Comment by dirk-thomas** [comment]
@BrannonKing Can you please try the patch from #109 which should fix the memory leak described in #74 and report back.

**Pull request closed without merging**